hw/ip/otbn/dv/otbnsim/sim/isa.py

```diff
@@ -162,8 +162,109 @@ def logical_byte_shift(value: int, shift_type: int, shift_bytes: int) -> int:
     return shifted & mask256
 
 
+def logical_bit_shift(value: int, size: int, shift_type: int, shift_bits: int) -> int:
+    '''Logical shift value by shift_bits to the left or right.
+
+    value should be an unsigned `size`-bit value. shift_type should be 0 (shift
+    left) or 1 (shift_right), matching the encoding of the big number
+    instructions. shift_bits should be a non-negative number of bits to shift
+    by.
+
+    Returns an unsigned `size`-bit value, truncating on an overflowing left shift.
+    '''
+    mask = (1 << size) - 1
+    assert 0 <= value <= mask
+    assert 0 <= shift_type <= 1
+    assert 0 <= shift_bits
+
+    shifted = value << shift_bits if shift_type == 0 else value >> shift_bits
+    return shifted & mask
+
+
 def extract_quarter_word(value: int, qwsel: int) -> int:
     '''Extract a 64-bit quarter word from a 256-bit value.'''
     assert 0 <= value < (1 << 256)
     assert 0 <= qwsel <= 3
     return (value >> (qwsel * 64)) & ((1 << 64) - 1)
+
+
+def extract_sub_word(value: int, size: int, index: int) -> int:
+    '''Extract a `size`-bit word at index `index` from a 256-bit value and
+    interprets it as unsigned integer'''
+    assert 0 <= value < (1 << 256)
+    assert 0 <= index <= 256 // size
+    return (value >> (index * size)) & ((1 << size) - 1)
+
+
+def lower_d_bits(value: int, d: int) -> int:
+    '''Extracts the lower d bits of the value.'''
+    return value & ((1 << d) - 1)
+
+
+def upper_d_bits(value: int, d: int) -> int:
+    '''Extracts the upper d bits of the value and shifts them down by d.'''
+    return (value & (((1 << d) - 1) << d)) >> d
+
+
+def montgomery_mul(a_, b_, q, R, size):
```

Comment:
What's the underscore in `a_` and `b_` for?

Reply:
To designate that a & b are in Montgomery space. But can be dropped
```suggestion
def montgomery_mul(a, b, q, R, size):
```